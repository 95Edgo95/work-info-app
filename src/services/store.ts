import createSagaMiddleware, {SagaMiddleware} from "redux-saga";
import {applyMiddleware, compose, createStore} from "redux";
import reducer from "services/reducer";
import saga from "services/saga";
import {fromJS} from "immutable";

let store: any;

const initialState: any = fromJS({});

const sagaMiddleware: SagaMiddleware<any> = createSagaMiddleware();

if (process.env && process.env.NODE_ENV !== "production") {
    const composeEnhancers: any = (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
    store = createStore(reducer, initialState, composeEnhancers(applyMiddleware(sagaMiddleware)));
} else {
    store = createStore(reducer, initialState, compose(applyMiddleware(sagaMiddleware)));
}

sagaMiddleware.run(saga);

export default store;
