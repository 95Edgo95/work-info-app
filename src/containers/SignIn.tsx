import {mapDispatchToProps, mapStateToProps} from "helpers/ReduxHelper";
import {attemptSignIn} from "store/user/UserActionCreators";
import {ISelectorVariables} from "services/selector";
import {RouteComponentProps} from "react-router";
import Button from "components/Button";
import Input from "components/Input";
import {connect} from "react-redux";
import * as React from "react";

import * as styles from "assets/css/signIn.css";

interface ISignInState {
  username: string;
  password: string;
  errors: string[];
}

interface ISignInReduxProps {
  isLoggedIn: boolean;
  errorFields: any;
}

interface ISignInDispatchProps {
  attemptSignIn: (username: string, password: string) => void;
}

interface ISignInProps extends ISignInReduxProps, ISignInDispatchProps, RouteComponentProps {}

class SignIn extends React.Component<ISignInProps, ISignInState> {
  state: ISignInState = {
    username: "",
    password: "",
    errors: [],
  };

  handleChange = ({currentTarget: {name, value}}: React.ChangeEvent<HTMLInputElement>) => {
    const newState: ISignInState = {...this.state};
    newState[name] = value;

    this.updateErrors(name, newState);
  };

  handleSubmit = (event: React.FormEvent) => {
    event.preventDefault();
    event.stopPropagation();

    const {username, password} = this.state;
    const {attemptSignIn} = this.props;

    attemptSignIn(username, password);
  };

  updateErrors(updatedKey: string, nextState: any): void {
    const {errors} = this.state;

    if (errors.includes(updatedKey)) {
      const nextErrors: string[] = [...errors];
      const index: number = errors.indexOf(updatedKey);
      nextErrors.splice(index, 1);

      this.setState({...nextState, errors: nextErrors});
    } else {
      this.setState(nextState);
    }
  }

  componentDidMount(): void {
    const {isLoggedIn, history} = this.props;

    if (isLoggedIn) {
      history.push("/");
    }
  }

  componentDidUpdate(prevProps: ISignInProps): void {
    const {isLoggedIn, history, errorFields} = this.props;

    if (!prevProps.isLoggedIn && isLoggedIn) {
      history.push("/");
    }

    if (errorFields.size > 0 && prevProps.errorFields.size === 0) {
      this.setState({errors: errorFields.toJS()});
    }
  }

  render(): JSX.Element {
    const {username, password, errors} = this.state;

    return (
      <div className={styles.wrapper}>
        <form onSubmit={this.handleSubmit} className={styles.form}>
          <div className={styles.inputWrapper}>
            <Input
              isValid={username.length > 0 && !errors.includes("username")}
              hasError={errors.includes("username")}
              onChange={this.handleChange}
              placeholder="Username"
              value={username}
              name="username"
              type="text"
            />
          </div>
          <div className={styles.inputWrapper}>
            <Input
              isValid={password.length > 0 && !errors.includes("password")}
              hasError={errors.includes("password")}
              onChange={this.handleChange}
              placeholder="Password"
              value={password}
              type="password"
              name="password"
            />
          </div>
          <div className={styles.actionBtnWrapper}>
            <Button type="submit">
              Sign In
            </Button>
          </div>
        </form>
      </div>
    );
  }
}

const selectorVariables: ISelectorVariables = {
  user: {
    errorFields: true,
    isLoggedIn: true
  }
};

const actions: any = {attemptSignIn};

export default connect<ISignInReduxProps, ISignInDispatchProps, any>(mapStateToProps(selectorVariables), mapDispatchToProps(actions))(SignIn);
