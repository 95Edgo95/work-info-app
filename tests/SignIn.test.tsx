import {BrowserRouter} from "react-router-dom";
import SignIn from "../src/containers/SignIn";
import store from "../__mocks__/storeMock";
import {Provider} from "react-redux";
import { mount } from "enzyme";
import * as React from "react";

describe('SignIn', () => {
  it('should render correctly', () => {
    const component = mount(
      <Provider store={store}>
        <BrowserRouter>
          <SignIn />
        </BrowserRouter>
      </Provider>
    );

    expect(component).toMatchSnapshot();
    component.unmount();
  });
});
