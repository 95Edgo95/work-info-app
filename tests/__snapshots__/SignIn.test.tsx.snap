// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn should render correctly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(SignIn)>
        <SignIn
          attemptSignIn={[Function]}
          errorFields={Immutable.List []}
          isLoggedIn={false}
        >
          <div>
            <form
              onSubmit={[Function]}
            >
              <div>
                <Input
                  hasError={false}
                  isValid={false}
                  name="username"
                  onChange={[Function]}
                  placeholder="Username"
                  type="text"
                  value=""
                >
                  <div>
                    <div
                      className=""
                    >
                      <input
                        name="username"
                        onChange={[Function]}
                        placeholder="Username"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </Input>
              </div>
              <div>
                <Input
                  hasError={false}
                  isValid={false}
                  name="password"
                  onChange={[Function]}
                  placeholder="Password"
                  type="password"
                  value=""
                >
                  <div>
                    <div
                      className=""
                    >
                      <input
                        name="password"
                        onChange={[Function]}
                        placeholder="Password"
                        type="password"
                        value=""
                      />
                    </div>
                  </div>
                </Input>
              </div>
              <div>
                <Button
                  type="submit"
                >
                  <button
                    className=""
                    type="submit"
                  >
                    Sign In
                  </button>
                </Button>
              </div>
            </form>
          </div>
        </SignIn>
      </Connect(SignIn)>
    </Router>
  </BrowserRouter>
</Provider>
`;
